File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        map <char, int> mp;
        int n = s.length(), m = t.length();
        if(n < m) return "";
        for(int i = 0 ; i < m ; i++) mp[t[i]]++;
        int l = 0, r = 0, cnt = 0, len = 1e9, si = -1;

        while(r < n){
            if(mp[s[r]] > 0) cnt++;
            mp[s[r]]--;
            while(cnt == m){
                mp[s[l]]++;
                if(mp[s[l]] > 0) cnt--;
                if(r-l+1 < len){
                    len = r-l+1;
                    si = l;
                }
                l++;
            }
            r++;
        }   
        if(len == 1e9) return "";
        return s.substr(si, len);
    }
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        from collections import defaultdict

        mp = defaultdict(int)
        n, m = len(s), len(t)
        if n < m:
            return ""

        # build frequency map for t
        for ch in t:
            mp[ch] += 1

        l = 0
        r = 0
        cnt = 0
        INF = 10**9
        length = INF
        si = -1

        while r < n:
            if mp[s[r]] > 0:
                cnt += 1
            mp[s[r]] -= 1

            while cnt == m:
                mp[s[l]] += 1
                if mp[s[l]] > 0:
                    cnt -= 1

                if r - l + 1 < length:
                    length = r - l + 1
                    si = l

                l += 1

            r += 1

        if length == INF:
            return ""
        return s[si: si + length]

        
__import__("atexit").register(lambda: open("display_runtime.txt", "w").write("000"))